Refactor video encoders to be more generic and self-contained

laundmo opened the pull request:
Hi there,

This is my current state regarding our conversation in
- #21

I've refactored it a bit more than I thought, but I think its for the better.
So far I've only tested nvenc.
It works, but I occasionally get 1 frame of `Could not send video frame at: 15221298411101. Channel full.` and consistently get `Error while draining nvenc encoder during drop: FFmpeg(ffmpeg::Error(541478725: ))`

Nevertheless, it seems to work and spits out a mp4 file.

Its not done yet, as I haven't done the main thing I wanted to: Allow an easy of getting raw frames.

But the groundwork is there.

laundmo commented on src/encoders/opus_encoder.rs:
```diff
@@ -71,7 +71,7 @@ impl AudioEncoder for OpusEncoder {
             let total_samples = raw_frame.samples.len();
 
-            if total_samples % n_channels != 0 {
+            if !total_samples.is_multiple_of(n_channels) {
```
That ones entirely `clippy` - it said "manual implementation of is_multiple_of" or something

laundmo commented on src/encoders/image_encoder.rs:
```diff
@@ -0,0 +1,38 @@
+use crate::VideoEncoder;
+use crossbeam::channel::{bounded, Receiver, Sender};
+
+struct ImageEncoder {
+    image_sender: Sender<image::RgbaImage>,
+    image_receiver: Receiver<image::RgbaImage>,
+}
```
Oh this just... isn't done.

I was naming it that because i plan to have it produce images in `image` crate format, so users don't have to mess with pixel formats and YUV conversion (currently, pipewire is allowed to produce 2 formats of YUV streams, not ideal for most usecases which are directly consuming frames)

laundmo replied:
I think part of this is a somewhat unclear communication on my part. I don't, in fact, want raw frames the way pipewire sees them (file descriptor to the memory) - i want "raw" images as in Vec<u8> of pixels i can work with or pass to other things. In my case, pattern detection and maybe even OCR. (i might, once this is done, integrate waycap into rustautogui as that's lacking a wayland backend)

laundmo commented on src/encoders/nvenc_encoder.rs:
```diff
@@ -386,6 +427,7 @@ impl NvencEncoder {
         Ok(())
     }
 
+    /// Set cuda and egl context to current thread
```
Yup, it used to also do egl but i moved that after some issues with egl context. not a fan of that whole thing, maybe i'll think of a better way

laundmo commented on src/encoders/nvenc_encoder.rs:
```diff
@@ -45,157 +48,168 @@ pub struct NvencEncoder {
         Ok(())
     }
 
+    fn thread_setup(&mut self) -> Result<()> {
```
that was the plan but it started producing viewable mp4 files without it so i thought "eh, good enough to get someone elses eyes on this" and forgot to remove it

laundmo commented on src/lib.rs:
```diff
@@ -167,57 +166,36 @@ impl Capture {
+                    stream_node,
+                    video_ready_pw,
+                    audio_ready_pw,
+                    matches!(encoder_type, VideoEncoderType::H264Nvenc),
```
aye, its just a bit shorter. part of my attempt to clean up the giant Capturer::new

laundmo commented on src/lib.rs:
```diff
@@ -442,153 +452,77 @@ impl Capture {
     }
 }
 
 #[allow(clippy::too_many_arguments)]
```
maybe, nvenc and egl contexts were giving me trouble so i wanted to mostly leave things as is before i break it *more*.

i'll try removing it, i assume you're more aware of how these work than me

laundmo commented on src/types/config.rs:
```diff
@@ -1,15 +1,15 @@
-#[derive(Debug)]
+#[derive(Debug, Clone, Copy)]
```
Theres absolutely no reason not to make 0 to 1 byte sized types copy/clone.

a reference might even be less efficient, since a reference is always at least of size `usize`. Instead, copying allows the compiler to often even completely optimize out enums like this.

do we need it? no. does it make everyones life easier, and possibly the whole thing compile better? yes.

laundmo commented:
> so this makes it seem like we try to pop one extra frame during drop which causes the error

Yeahhhh not a fan of the Drop logic right now. I think i need to list everything out at some point and just... re-do it. i feel like we should need less logic called on Drop, and also in less nested ways.

honestly, i'm of the opinion that if someone drops the encoder before calling .finish() on it or whatever, its their fault if they loose some frames. so maybe the drain() calls could go completely.

laundmo commented:
*it just needs to use the nvenc variants of the pipewire config, thats all. totally didn't spend like, 5 hours bashing my head against this*

Adonca2203 commented:
yeah you think i came up with those magic number by myself lol I was stuck on that for a good while

laundmo commented:
Asking Pipewire directly for the raw frame makes much more sense than what i was attempting. part of the issue is that DMA Buffers live on the GPU and so i would have to make a separate version for nvenv and vaapi and figure out the correct ffmpeg filters. getting it directly is so much simpler and i'm gonna try that right now.

laundmo commented:
That is SO MUCH nicer! Thank you!

laundmo commented on src/lib.rs:
```diff
@@ -218,11 +218,11 @@ impl<V: VideoEncoder> Capture<V> {
                     reso_sender,
                     frame_tx,
                     pw_recv,
+                    V::get_spa_definition()?,
```
What, calling a static trait method on a generic?

Adonca2203 replied:
yeah im still very much a rust noob and feel like I have too much OOP in the brain when using this when I know I should be thinking functional

laundmo commented:
Ready for review/merge, at least from my end.

With this far more enormous chunk of effort thank i expected out of the way, i hope i'm done.

Notably: I changed how the thread setup works in its entirety, added a second "decoder" which just gives the user the RawVideoFrame reciever, cleaned up docs and especially made the few doctests work, so `cargo test` now works.

And probably some more stuff i'm not remembering right now.
